Improve error message when typo is made in format!

scrabsha opened the pull request:
The expansion of the format! built-in macro is roughly done in two steps:
  - the format expression is parsed, the arguments are parsed,
  - the format expression is checked to be a string literal, code is expanded.

The problem is that the expression parser can eat too much tokens, which invalidates the parsing of the next format arguments. As the format expression check happens next, the error emitted concerns the format arguments, whereas the problem is about the format expression.

This PR contains two commits. The first one actually checks that the formatting expression is a string literal before raising any error about the formatting arguments, and the second one contains some simple heuristics which allow to suggest, when the format expression is followed by a dot instead of a comma, to suggest to replace the dot with a comma.

This pull request should fix #75492.

Note: this is my first non-doc contribution to the rust ecosystem. Feel free to make any comment about my code, or whatever. I'll be very happy to fix it :)

ecstatic-morse commented:
r? @petrochenkov 

I thought highfive had some heuristic based on who has touched the file in question.

petrochenkov commented:
So, this PR looks significantly more complicated than it needs to be.

There are two orthogonal issues here:
- the format expression is parsed over-eagerly
- recovery for confusable separators `.` and `,` is not performed

I suggest addressing them independently starting from the first one.
Simple implementation for it:
- Before parsing the format expression use one token `look_ahead`
    - if the next token is a string literal make a format expression out of it, `bump()` and proceed parsing other arguments
    - otherwise use `parse_expr` to make a format expression, and proceed parsing other arguments

scileo commented:
Thanks for your feedback, I'm working on it :)

Side question: should I open a new PR, or is it okay to modify this one?

petrochenkov commented:
@scileo 
Whichever is more convenient, both ways are ok.

scrabsha committed:
Use string literal directly when available in format

Previous implementation used the `Parser::parse_expr` function in order
to extract the format expression. If the first comma following the
format expression was mistakenly replaced with a dot, then the next
format expression was eaten by the function, because it looked as a
syntactically valid expression, which resulted in incorrectly spanned
error messages.

The way the format expression is exctracted is changed: we first look at
the first available token in the first argument supplied to the
`format!` macro call. If it is a string literal, then it is promoted as
a format expression immediatly, otherwise we fall back to the original
`parse_expr`-related method.

This allows us to ensure that the parser won't consume too much tokens
when a typo is made.

A test has been created so that it is ensured that the issue is properly
fixed.

scileo commented:
@petrochenkov 
I have applied all the changes you suggested. Emitted errors are way nicer now!

There are a few changes that may need to be discussed:

  - when a string literal is detected, but the string literal itself is incorrect (eg: [this branch][incorrect-strlit]), a fake string literal is generated. I came up with this solution by comparing what `parse_expr` emits in this case. However, its `symbol` field slightly differs from what `parse_expr` creates: opening and closing quotes are not included. Still all pass, but I am not sure if this is correct.
  - in order to create the format expression, I ended up using roughly the same code as [`Parser::mk_expr`]. Is it correct?
  - your suggestion does not detect cases when the format expression is not a string literal, such as `format!(concat!("{", "}"). 42)`. This looks acceptable to me, as non-string literal format expression is pretty rare.
  - should we also handle raw strings too?
  - you talked earlier about error recovery. Is it needed now?

[incorrect-strlit]: https://github.com/rust-lang/rust/pull/75779/files#diff-69112157279b9e4a1dd02ce18d607250R151-R157
[`Parser::mk_expr`]: https://github.com/rust-lang/rust/blob/d8424f6b426f91ae39dbeacd631a82aad5d733f4/src/librustc_parse/parser/expr.rs#L2286-L2288

petrochenkov commented on src/librustc_builtin_macros/format.rs:
```diff
@@ -135,7 +136,43 @@ fn parse_args<'a>(
+
+            p.bump();
+
+            P(ast::Expr { span, kind, attrs, tokens, id })
```
You could use `p.parse_literal_maybe_minus()` to avoid implementing this logic by hand. (It's not currently public but it's ok to make it public.)

petrochenkov commented on src/librustc_builtin_macros/format.rs:
```diff
@@ -135,7 +136,43 @@ fn parse_args<'a>(
-    let fmtstr = p.parse_expr()?;
+    let first_token = &p.token;
+    let fmtstr = match first_token.kind {
+        token::TokenKind::Literal(lit_token @ token::Lit { kind: token::LitKind::Str, .. }) => {
```
```suggestion
        token::TokenKind::Literal(lit_token @ token::Lit { kind: token::Str | token::StrRaw, .. }) => {
```

petrochenkov commented:
>you talked earlier about error recovery. Is it needed now?

Yes, but it's better to do it in a separate PR later.
We basically need to treat `.` as `,` (but with a non-fatal error) if we encounter it as a separator in the `format!` arguments list.
See the use of `similar_tokens` in `parse_seq_to_before_tokens` for an example.

scileo commented:
> > you talked earlier about error recovery. Is it needed now?
> 
> Yes, but it's better to do it in a separate PR later.
> We basically need to treat `.` as `,` (but with a non-fatal error) if we encounter it as a separator in the `format!` arguments list.
> See the use of `similar_tokens` in `parse_seq_to_before_tokens` for an example.

Thanks for the clarification. My initial understandings about how to do it were wrong. I'll work on it once this PR is merged.

petrochenkov commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit 1001c76c6c449e27d90409ff5cd8865d510f5341 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 1001c76c6c449e27d90409ff5cd8865d510f5341 -->
<!-- homu: {"type":"Approved","sha":"1001c76c6c449e27d90409ff5cd8865d510f5341","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 1001c76c6c449e27d90409ff5cd8865d510f5341 with merge 1cb116072f07171b4a3d7a5cd61a8c51514445c7...
<!-- homu: {"type":"BuildStarted","head_sha":"1001c76c6c449e27d90409ff5cd8865d510f5341","merge_sha":"1cb116072f07171b4a3d7a5cd61a8c51514445c7"} -->

petrochenkov commented:
@bors retry
(Prioritizing https://github.com/rust-lang/rust/pull/74862)

bors commented:
:hourglass: Testing commit 1001c76c6c449e27d90409ff5cd8865d510f5341 with merge 1a5c36a0e53086af131b47592bbaaea4cf254c62...
<!-- homu: {"type":"BuildStarted","head_sha":"1001c76c6c449e27d90409ff5cd8865d510f5341","merge_sha":"1a5c36a0e53086af131b47592bbaaea4cf254c62"} -->

rust-log-analyzer commented:
Your PR [failed](https://github.com/rust-lang-ci/rust/runs/1047998285) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/1a5c36a0e53086af131b47592bbaaea4cf254c62/checks/1047998285/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/1a5c36a0e53086af131b47592bbaaea4cf254c62/checks/1047998285/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[group]Run exit 1
exit 1
shell: /bin/bash --noprofile --norc -e -o pipefail {0}
##[endgroup]
##[error]Process completed with exit code 1.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:hourglass: Testing commit 1001c76c6c449e27d90409ff5cd8865d510f5341 with merge c57047b5fd90f5aaafb6028350ba44356c20f56f...
<!-- homu: {"type":"BuildStarted","head_sha":"1001c76c6c449e27d90409ff5cd8865d510f5341","merge_sha":"c57047b5fd90f5aaafb6028350ba44356c20f56f"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #74862) made this pull request unmergeable. Please resolve the merge conflicts.

Aaron1011 commented:
@bors retry

Aaron1011 commented:
@bors r-

petrochenkov commented:
Needs a rebase.
(Also, it's preferable to squash the commits into one.)

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit f6d18db402cfebcf5c1bdc8e730f3923b3bb0421 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov f6d18db402cfebcf5c1bdc8e730f3923b3bb0421 -->
<!-- homu: {"type":"Approved","sha":"f6d18db402cfebcf5c1bdc8e730f3923b3bb0421","approver":"petrochenkov"} -->

scileo commented:
Thank you very much for your help @petrochenkov!

bors commented:
:hourglass: Testing commit f6d18db402cfebcf5c1bdc8e730f3923b3bb0421 with merge 36b0d7e25769e88fec85e1d073196065a7f2d7c4...
<!-- homu: {"type":"BuildStarted","head_sha":"f6d18db402cfebcf5c1bdc8e730f3923b3bb0421","merge_sha":"36b0d7e25769e88fec85e1d073196065a7f2d7c4"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1048706069), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36318)
Approved by: petrochenkov
Pushing 36b0d7e25769e88fec85e1d073196065a7f2d7c4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36318","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1048706069"},"merge_sha":"36b0d7e25769e88fec85e1d073196065a7f2d7c4"} -->
